### PR TITLE
Change Item Hierarchy

### PR DESCRIPTION
**Pull request by querdenker2k**
Changed the Item hierarchy and removed the generation of the specific Items. Just the JRuleItems will be generated.
A XXXGroupItem will extend the XXXItem.

### CODE REVIEW

**Comment by querdenker2k** [comment]
@seaside1 Important things i've done
- I removed the XYValue and RGBValue, just HSB exist. I could add RGB again if required
- Created InternalItems for every ItemType
- Added a lot of tests for the Items
- New JRuleItemName Class for referencing the Itemnames for the rule conditions. Not possible with JRuleItems anymore, because there is no static value anymore. First I tried to create an enum for this, but failed, because i don't see any option for testing

**Comment by querdenker2k** [comment]
Using this in my productive environment now.
@seaside1 I know this is again a huge mr. What do you need for aprooving?

**Comment by seime** [comment]
Thanks @querdenker2k . I started looking around yesterday, but as this PR is huge I would need some more time to update my rules and look at the code.

**Comment by querdenker2k** [comment]
But i am having problems getting the integration test stable

**Comment by seaside1** [comment]
Sorry for not commenting earlier. Will also start to review and test it out.

**Comment by querdenker2k** [comment]
the splitted loglines fix for the integration test seems to work now

**Comment by querdenker2k** [comment]
There seems to be a problem sending String commands (in the right format) to any item-type.
I will change the state types in the JRuleItemEvent to the JRuleValue types 
and allow to send any kind of value to any item. Why?
I have rules like this
```
    @JRuleName("light proxy to proxy")
    @JRuleWhenItemReceivedCommand(item = JRuleItemName.G_Light_Proxy, memberOf = true)
    @JRuleTag({Tags.LIGHT})
    public void lightProxyToLight(JRuleItemEvent event) {
        if (JRuleItem.forName(event.getMemberName()).isGroup()) {
            return;
        }
        var lightName = event.getMemberName().substring(12);
        logInfo("forwarding proxy to real item");
        var itemLight = JRuleItem.forName(String.format("Light_%s", lightName));
        itemLight.sendCommand(event.getState().getValueAsInteger());
    }
```
I don't know what kind of Item Light_XXX is (could be Switch or Dimmer or Color)
The option would be to check the instance of itemLight, but then I have to check the instance of event.getState() as well.

**Comment by querdenker2k** [comment]
I changed the class of the state property of the JRuleItemEvent to JRuleValue now. Shall I do this in this pull request as well?
@seime @seaside1

**Comment by querdenker2k** [comment]
push

**Comment by seaside1** [comment]
Yes you might as well I recon (JRuleItemEvent / JRuleValue). I'm running testing this in my prod instance now. I'll report back soon.
How is it going for you @seime ? It seems my rules are not affected to any larger extent.
Regarding naming, I guess we have JRuleItems, so JRuleItemNames would follow that pattern.

**Comment by seime** [comment]
I've been unable to do any review yet, hopefully will find some time within a week

**Comment by seaside1** [comment]
I'm running it without issue.
- Maybe we should add back support for RGB for ColorItems, that could be done at a later stage
-  Some of the the changes might be a bit subjective if they are actually making things clearer or better, but I don't see any issues with it right now and overall it's an improvement. 
- I like the added tests.
I think we should merge this one first before we start merging the other PR:s, Since it such a big PR I think it's good if Seime can give a review as well.

**Review by seime — COMMENTED** [review]
I've added a few preliminary comments. Lots of code changes, will take many more hours to through it all. Findings so far:

- Some missing sendCommand/postUpdates
- A bit inconsistent when JRuleXXXItem has constants for some of the textual representation of enum values and when not
- We should introduce JRuleItem.postUndef() to set items to UNDEF
- Similar JRuleItem.refresh() to send command `RefreshType`
- `sendCommand(String)` and `postUpdate(String)` should be reserved for `String` items only, and the existing should be renamed to `sendUncheckedCommand` to signal to the user that he/she basically overrides type checking and that the consequence is blowup in production
- Some confusing sendCommand(boolean) around
- We should add Javadoc to user facing methods
- Could we merge `JRuleItems` and `JRuleItemNames` by adding an inner static class to `JRuleItems`? We could then refer to item names like `JRuleItems.Name.MyItemName`

Awaiting some response from @querdenker2k  and @seaside1  before continuing code review

**Comment by querdenker2k** [comment]
> I've added a few preliminary comments. Lots of code changes, will take many more hours to through it all. Findings so far:
> 
> * Some missing `sendCommand`/ `postUpdate`
> * A bit inconsistent when `JRuleXXXItem` has constants for some of the textual representation of enum values and when not
> * We should introduce `JRuleItem.postUndef()` to set items to `UNDEF`
> * Similar `JRuleItem.refresh()` to send command `RefreshType`
> * `sendCommand(String)` and `postUpdate(String)` should be reserved for `String` items only, and the existing should be renamed to `sendUncheckedCommand` to signal to the user that he/she basically overrides type checking and that the consequence is blowup in production
> * Some confusing sendCommand(boolean) around
> * We should add Javadoc to user facing methods
> * Could we merge `JRuleItems` and `JRuleItemNames` by adding an inner static class to `JRuleItems`? We could then refer to item names like `JRuleItems.Name.MyItemName`
> 
> Awaiting some response from @querdenker2k and @seaside1 before continuing code review

1. Which sendCommand/postUpdate method you mean? I am using this MR for 3 weeks now
2. I will add the missing ones i find.
3. This is realized via postUpdate(null). I had another branch which cleaned up the value stuff, merged this into this branch now
4. I saw this type in the oh code but didn't when to use it so i didn't implemented it. Is it realy used?
5. I don't understand how the sendCommand() with String parameter worked before because in the oh Items Stringtype is not registered as supported command/state.
6. I added the boolean methods to do something like sendCommand(state.equals("foo")) -> true is always on/open/up/100%
7. I will add more javadoc
8. I would not like to merge this because the generated class is getting bigger and bigger. Currently I have ~2700 Items.

File: src/main/java/org/openhab/automation/jrule/rules/value/JRuleDateTimeValue.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            "yyyy-MM-dd'T'HH:mm[:ss[.SSSSSSSSS][.SSSSSSSS][.SSSSSSS][.SSSSSS][.SSSSS][.SSSS][.SSS][.SS][.S]]");
+    private static final DateTimeFormatter PARSER_TZ = DateTimeFormatter.ofPattern(
+            "yyyy-MM-dd'T'HH:mm[:ss[.SSSSSSSSS][.SSSSSSSS][.SSSSSSS][.SSSSSS][.SSSSS][.SSSS][.SSS][.SS][.S]]z");
+    private static final DateTimeFormatter PARSER_TZ_RFC = DateTimeFormatter.ofPattern(
+            "yyyy-MM-dd'T'HH:mm[:ss[.SSSSSSSSS][.SSSSSSSS][.SSSSSSS][.SSSSSS][.SSSSS][.SSSS][.SSS][.SS][.S]]Z");
+    private static final DateTimeFormatter PARSER_TZ_ISO = DateTimeFormatter.ofPattern(
+            "yyyy-MM-dd'T'HH:mm[:ss[.SSSSSSSSS][.SSSSSSSS][.SSSSSSS][.SSSSSS][.SSSSS][.SSSS][.SSS][.SS][.S]]X");
+    private static final DateTimeFormatter FORMATTER_TZ_RFC = DateTimeFormatter
+            .ofPattern("yyyy-MM-dd'T'HH:mm[:ss[.SSSSSSSSS]]Z");
+
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/rules/value/JRuleDateTimeValue.java:43]: This is removed with the merge (of my other branch). The JRule values will just wrap up the oh types

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalContactItem.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        super(name, label, type, id);
+    }
+
+    public void sendCommand(boolean command) {
+        JRuleEventHandler.get().sendCommand(getName(), JRuleOpenClosedValue.valueOf(command));
+    }
+
+    public void postUpdate(boolean command) {
+        JRuleEventHandler.get().postUpdate(getName(), JRuleOpenClosedValue.valueOf(command));
+    }
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalContactItem.java:36]: JRuleOpenClosedValue is the default type, see the generic type and JRuleItem

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalDimmerItem.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        JRuleEventHandler.get().postUpdate(name, new JRulePercentValue(value));
+    }
+
+    public void sendCommand(boolean command) {
+        JRuleEventHandler.get().sendCommand(getName(), new JRulePercentValue(command));
+    }
+
+    public void postUpdate(boolean command) {
+        JRuleEventHandler.get().postUpdate(getName(), new JRulePercentValue(command));
+    }
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalDimmerItem.java:58]: I would like to do this sendCommand(value.equals("turn it on"));

File: src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ */
+public class JRulePercentValue extends JRuleDecimalValue {
+    public JRulePercentValue(int value) {
+        super(value);
+    }
+
+    public JRulePercentValue(double value) {
+        super((int) Math.round(value + 0.5));
+    }
+
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java:32]: within the merged branch this is handled by the oh type

File: src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    public JRulePercentValue(String value) {
+        super(value);
+    }
+
+    public JRulePercentValue(boolean value) {
+        super(value ? 100 : 0);
+    }
+
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java:40]: as above suggested why not ;)

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalImageItem.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.openhab.automation.jrule.items.JRuleImageItem;
+import org.openhab.automation.jrule.rules.value.JRuleRawValue;
+
+/**
+ * The {@link JRuleInternalImageItem} Items
+ *
+ * @author Arne Seime - Initial contribution
+ */
+public class JRuleInternalImageItem extends JRuleInternalItem<JRuleRawValue> implements JRuleImageItem {
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalImageItem.java:23]: JRuleRawValue is the only accepted type

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalLocationItem.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.openhab.automation.jrule.items.JRuleLocationItem;
+import org.openhab.automation.jrule.rules.value.JRulePointValue;
+
+/**
+ * The {@link JRuleInternalLocationItem} Items
+ *
+ * @author Arne Seime - Initial contribution
+ */
+public class JRuleInternalLocationItem extends JRuleInternalItem<JRulePointValue> implements JRuleLocationItem {
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalLocationItem.java:23]: see JRuleItem. its supported

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalNumberItem.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+import org.openhab.automation.jrule.items.JRuleNumberItem;
+import org.openhab.automation.jrule.rules.value.JRuleDecimalValue;
+import org.openhab.core.library.types.DecimalType;
+
+/**
+ * The {@link JRuleInternalNumberItem} Items
+ *
+ * @author Joseph (Seaside) Hagberg - Initial contribution
+ */
+public class JRuleInternalNumberItem extends JRuleInternalItem<JRuleDecimalValue> implements JRuleNumberItem {
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalNumberItem.java:28]: i hope this is handled by the wrapped oh type

File: src/main/java/org/openhab/automation/jrule/items/JRuleItem.java [hunk 9]
```diff
@@ -17,59 +17,82 @@
...
 
-    public abstract String getLabel();
+    default void postUpdate(T state) {
+        JRuleEventHandler.get().postUpdate(getName(), state);
+    }
 
-    public abstract String getType();
+    default void sendCommand(String state) {
+        JRuleEventHandler.get().sendCommand(getName(), new JRuleStringValue(state));
+    }
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/items/JRuleItem.java:65]: you're right, was already removed in the other branch

File: src/main/java/org/openhab/automation/jrule/items/JRuleItem.java [hunk 10]
```diff
@@ -17,59 +17,82 @@
...
 
-    public abstract String getId();
+    default void postUpdate(String state) {
+        JRuleEventHandler.get().postUpdate(getName(), new JRuleStringValue(state));
+    }
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/items/JRuleItem.java:69]: see above

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalPlayerItem.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        JRuleEventHandler.get().postUpdate(name, state);
+    }
+
+    public void sendCommand(boolean command) {
+        JRuleEventHandler.get().sendCommand(getName(), JRulePlayPauseValue.valueOf(command));
+    }
+
+    public void postUpdate(boolean command) {
+        JRuleEventHandler.get().postUpdate(getName(), JRulePlayPauseValue.valueOf(command));
+    }
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalPlayerItem.java:44]: see above

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalPlayerItem.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+import org.openhab.automation.jrule.internal.handler.JRuleEventHandler;
+import org.openhab.automation.jrule.items.JRulePlayerItem;
+import org.openhab.automation.jrule.rules.value.JRulePlayPauseValue;
+
+/**
+ * The {@link JRuleInternalPlayerItem} Items
+ *
+ * @author Joseph (Seaside) Hagberg - Initial contribution
+ */
+public class JRuleInternalPlayerItem extends JRuleInternalItem<JRulePlayPauseValue> implements JRulePlayerItem {
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalPlayerItem.java:24]: i allowed currently JRuleValue (in the merged branch)

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalRollershutterItem.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        JRuleEventHandler.get().postUpdate(name, new JRulePercentValue(value));
+    }
+
+    public void sendCommand(boolean command) {
+        JRuleEventHandler.get().sendCommand(getName(), JRuleUpDownValue.valueOf(command));
+    }
+
+    public void postUpdate(boolean command) {
+        JRuleEventHandler.get().postUpdate(getName(), JRuleUpDownValue.valueOf(command));
+    }
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalRollershutterItem.java:59]: i will add javadoc for this

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalImageItem.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (seime) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalImageItem.java:23]: But the default impl also adds a sendCommand which is not supported by openhab?

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalLocationItem.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (seime) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalLocationItem.java:23]: You're rigth

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalStringItem.java [hunk 14]
```diff
@@ -10,26 +10,19 @@
...
-    private final int value;
-
-    public JRulePercentType(int value) {
-        this.value = value;
-    }
-
-    public JRulePercentType(double value) {
-        this.value = (int) Math.round(value + 0.5);
-    }
+public class JRuleInternalStringItem extends JRuleInternalItem<JRuleStringValue> implements JRuleStringItem {
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalStringItem.java:23]: changed in merged branch

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalDimmerItem.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (seime) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalDimmerItem.java:58]: Can't you do that with `sendCommand(value.equals("turn it on") ? 100 : 0)`?

At least it should be added javadoc to the method to explain how it actually works

File: src/main/java/org/openhab/automation/jrule/items/JRuleContactItem.java [hunk 15]
```diff
@@ -12,36 +12,29 @@
...
+ * @author Robert Delbrück - Initial contribution
  */
-public abstract class JRuleContactItem extends JRuleItem implements JRuleContactTrigger {
-
-    protected JRuleContactItem(String itemName) {
-        super(itemName);
-    }
+public interface JRuleContactItem extends JRuleItem<JRuleOpenClosedValue> {
+    String OPEN = "OPEN";
+    String CLOSED = "CLOSED";
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/items/JRuleContactItem.java:27]: you're right, will change this

File: src/main/java/org/openhab/automation/jrule/items/JRulePlayerItem.java [hunk 16]
```diff
@@ -12,44 +12,29 @@
...
+ * @author Robert Delbrück - Initial contribution
  */
-public abstract class JRulePlayerItem extends JRuleItem implements JRulePlayerTrigger {
-
-    protected JRulePlayerItem(String itemName) {
-        super(itemName);
-    }
+public interface JRulePlayerItem extends JRuleItem<JRulePlayPauseValue> {
+    String PLAY = "PLAY";
+    String PAUSE = "PAUSE";
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/items/JRulePlayerItem.java:26]: will change this.

**Comment by querdenker2k** [comment]
will add javadocs in the evening.

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalDimmerItem.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalDimmerItem.java]: of course, but its longer ;)
javadoc is added

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalImageItem.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalImageItem.java:23]: thats right for now, actually all items allow all JRuleValues for easier usage (for me). the other option would be to again allow just the supported ones and add a "sendUnsafeCommand" and "postUnsafeUpdate" which accept all JRuleValues.

**Comment by seaside1** [comment]
What is the status now then?
Yes I agree that it can be a bit of fuzzy logic to have a for instance. JRuleContactItem.sendCommand(boolean value)
In that case I think it's better to use JRuleOpenCloseValue or similar.

**Comment by querdenker2k** [comment]
on my behalf, i will remove all sendCommand/postUpdate with boolean parameters.
what else?

**Comment by seime** [comment]
Values are now wrapped in a JRuleValue-class

File: src/main/java/org/openhab/automation/jrule/items/JRulePlayerItem.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (seime) [reply src/main/java/org/openhab/automation/jrule/items/JRulePlayerItem.java]: Missing sendCommand/postUpdate for the mentioned types

File: src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+import org.openhab.core.library.types.PercentType;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+
+/**
+ * The {@link JRulePercentValue}
+ *
+ * @author Joseph (Seaside) Hagberg - Initial contribution
+ */
+public class JRulePercentValue implements JRuleValue {
```
Comment (seime) [inline src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java:27]: Should this not extend JRuleDecimalValue to align with openhab internals?

**Review by seime — COMMENTED** [review]
sendCommand(boolean) and postUpdate(boolean) methods seems to be still present around, I think you wrote you'd remove them?

**Comment by seime** [comment]
It is not possible do a `postUpdate(null)` whenever there overloaded versions. Must introduce a fake cast to one of the methods ie `postUpdate((Date) null)` to achieve this

**Comment by seime** [comment]
Question: Ie getState().asStringValue() when state is null. Will getState() return null, or getState().asStringValue() (and others) return null?

**Comment by seime** [comment]
I also find it a bit painful to do runtime casts of ie JRuleItemEvent.getState()/getOldState(). Blows up at runtime / unit tests if not correct. Could we find a more typesafe solution?

**Comment by seime** [comment]
Should we have a separate JRuleQuantityTypeItem / internal class to avoid mixing these together? Now it is possible to send a quantity/unit to a non-quantity type item?

File: src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java (hunk 17 above)
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java:27]: done

**Comment by querdenker2k** [comment]
> sendCommand(boolean) and postUpdate(boolean) methods seems to be still present around, I think you wrote you'd remove them?

yes, i said i will do. but i don't want to do every small change "right now".
done now

**Comment by querdenker2k** [comment]
> It is not possible do a `postUpdate(null)` whenever there overloaded versions. Must introduce a fake cast to one of the methods ie `postUpdate((Date) null)` to achieve this

added extra method for this **postNullUpdate**

**Comment by querdenker2k** [comment]
> Question: Ie getState().asStringValue() when state is null. Will getState() return null, or getState().asStringValue() (and others) return null? If the first this now requires the rule code to be littered with null checks for getState().

done

**Comment by querdenker2k** [comment]
> I also find it a bit painful to do runtime casts of ie JRuleItemEvent.getState()/getOldState(). Blows up at runtime / unit tests if not correct. Could we find a more typesafe solution?

maybe we can do this later, this PR is getting bigger and bigger...

**Comment by querdenker2k** [comment]
> Should we have a separate JRuleQuantityTypeItem / internal class to avoid mixing these together? Now it is possible to send a quantity/unit to a non-quantity type item?

probably we should, but i would say in another PR.

**Comment by seime** [comment]
OK, could you just resolve the open conversations that indeed have been resolved, and keep the unresolved open to make it easy to create separate tasks?

Fully agreed that the PR is big and should be merged now. Great wort @querdenker2k !

Do you want to take a final look and merge if you're happy @seaside1 ?

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalContactItem.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalContactItem.java]: generics has been removed again

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalContactGroupItem.java [hunk 18]
```diff
@@ -10,48 +10,43 @@
...
-        return JRulePersistenceExtensions.historicState(itemName, timestamp, persistenceServiceId)
-                .map(s -> JRuleEventHandler.get().getOpenClosedValue(s));
+    public void sendCommand(boolean command) {
+        JRuleEventHandler.get().getGroupMemberNames(name, false)
+                .forEach(s -> JRuleEventHandler.get().sendCommand(s, JRuleOnOffValue.valueOf(command)));
+    }
+
+    public void postUpdate(boolean command) {
+        JRuleEventHandler.get().getGroupMemberNames(name, false)
+                .forEach(s -> JRuleEventHandler.get().postUpdate(s, JRuleOnOffValue.valueOf(command)));
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalContactGroupItem.java]: resolved as above

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalDimmerItem.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalDimmerItem.java]: removed boolean options for all item types, except SwitchItem

File: src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java]: removed

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalNumberItem.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalNumberItem.java]: enhancing the QuantityTypes will be a new merge-request

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalPlayerItem.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalPlayerItem.java]: added

File: src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalRollershutterItem.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalRollershutterItem.java]: removed boolean option

File: src/main/java/org/openhab/automation/jrule/items/JRulePlayerItem.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/items/JRulePlayerItem.java]: added

File: src/main/java/org/openhab/automation/jrule/items/JRuleItem.java [hunk 19]
```diff
@@ -17,59 +17,82 @@
...
 
-    public String getStateAsString() {
-        return JRuleEventHandler.get().getStringValue(itemName);
+    default <TD extends JRuleValue> TD getStateAs(Class<TD> type) {
+        return (TD) JRuleEventHandler.get().getValue(getName(), type);
     }
 
-    public String getName() {
-        return itemName;
+    default void sendCommand(T command) {
```
[the remark this answers is not in the document]
Comment (querdenker2k) [reply src/main/java/org/openhab/automation/jrule/items/JRuleItem.java]: done

**Comment by seaside1** [comment]
Yes looks good. Lets merge!

**Review by seaside1 — APPROVED** [review]
Approved!

**Pull request merged**